### PR TITLE
Create Excel example #121

### PR DESCRIPTION
**Pull request by markosimulak**
Hi Marcos,

I wrote these 2 excel examples. Please check them and tell me if there is something left to be added or I need to do some changes.

Cheers,
Marko Simulak

### CODE REVIEW

File: FileHelpers.Examples/Examples/13.Excel/10.ExcelCreateAndSave.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+            students[1] = new Student();
+            students[1].StudentNumber = 1;
+            students[1].FullName = "Steven Seagal";
+            students[1].Course = "Aikido";
+
+            students[2] = new Student();
+            students[2].StudentNumber = 2;
+            students[2].FullName = "Dennis Ritchie";
+            students[2].Course = "Programming";
```
Comment (ivandrofly) [inline FileHelpers.Examples/Examples/13.Excel/10.ExcelCreateAndSave.cs]: Maybe a factory method?

``` c#
    GetStudent(int num, string fullName, string course) =>
        new Student() { StudentNumber = num, FullName = fullName, Course = course } ;
```

Comment (markosimulak) [inline FileHelpers.Examples/Examples/13.Excel/10.ExcelCreateAndSave.cs]: Sure, I can do that. Looks cleaner

File: FileHelpers.Examples/Examples/13.Excel/10.ExcelCreateAndSave.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+using FileHelpers.ExcelNPOIStorage;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace ExamplesFx
+{
+    //-> Name: Create excel storage and save it.
+    //-> Runnable: true
```
Comment (mcavigelli) [inline FileHelpers.Examples/Examples/13.Excel/10.ExcelCreateAndSave.cs]: What does this annotation _Runnable: true_ mean? Also in other example file.

File: FileHelpers.Examples/Examples/13.Excel/10.ExcelCreateAndSave.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        }
+        //-> /File
+
+        
+        //-> File: CreateStudent.cs
+        /// <summary>
+        /// Create new student
+        /// </summary>
+        /// <returns>Student object</returns>
+        public static Student CreateStudent(int studentNumber, string fullName, string course)
```
Comment (mcavigelli) [inline FileHelpers.Examples/Examples/13.Excel/10.ExcelCreateAndSave.cs]: Method can be private.

File: FileHelpers.Examples/Examples/13.Excel/20.ExportAndEdit.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            // startRow = 2 & startColumn = 1 -> for skipping column header names
+            ExcelNPOIStorage storage = new ExcelNPOIStorage(typeof(Student), 2, 1);
+
+            // Set storage file name -> represents the excel file name we want to read
+            storage.FileName = "Students.xlsx";
+
+            // Read from excel file
+            Student[] students = storage.ExtractRecords() as Student[];
+
+            Console.WriteLine("\t\tStudents from file:\n");
```
Comment (mcavigelli) [inline FileHelpers.Examples/Examples/13.Excel/20.ExportAndEdit.cs]: I suggest you remove the _\n_ in both the example files.

_WriteLine()_ already writes a new line. _\n_ is not a proper new line on windows.
If you really need a robust new line that works
on all the operating systems, use Environment.NewLine.

File: FileHelpers.Examples/Examples/13.Excel/20.ExportAndEdit.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+            // This method will save out excel file
+            storage.InsertRecords(students);
+            Console.WriteLine("Changes saved.");
+
+            //-> /File
+        }
+
+
+        //-> File: Customer.cs
+        [DelimitedRecord("")]
```
Comment (mcavigelli) [inline FileHelpers.Examples/Examples/13.Excel/20.ExportAndEdit.cs]: This should probably be _File: Student.cs_

File: FileHelpers.Examples/Examples/ExamplesGenerator.autogen.cs [hunk 6]
```diff
@@ -8,849 +8,849 @@
```
Comment (mcavigelli) [inline FileHelpers.Examples/Examples/ExamplesGenerator.autogen.cs]: I guess this file is autogenerated :-)

I did not manage to generate again: I tried ExamplesGenerator.tt ->context-menu->Run Custom Tool.
Curious, about how it should be done.

File: FileHelpers.Examples/FileHelpers.Examples.csproj [hunk 7]
```diff
@@ -186,6 +188,14 @@
...
   <ItemGroup>
+    <ProjectReference Include="..\FileHelpers.ExcelNPOIStorage\FileHelpers.ExcelNPOIStorage.csproj">
+      <Project>{beb0d823-6a7e-47fd-a275-1b3314f8412d}</Project>
+      <Name>FileHelpers.ExcelNPOIStorage</Name>
+    </ProjectReference>
+    <ProjectReference Include="..\FileHelpers.ExcelStorage\FileHelpers.ExcelStorage.csproj">
+      <Project>{b3f2b976-73d5-49c7-beaa-bf62e233132a}</Project>
+      <Name>FileHelpers.ExcelStorage</Name>
+    </ProjectReference>
     <ProjectReference Include="..\FileHelpers\FileHelpers.csproj">
```
Comment (mcavigelli) [inline FileHelpers.Examples/FileHelpers.Examples.csproj]: I think the reference to Excel (COM) Storage is not required: please remove it.
The references should be kept as small as possible to see the real project
dependencies.

File: FileHelpers.ExcelStorage/ExcelStorage.cs [hunk 8]
```diff
@@ -215,8 +215,9 @@ private void SaveWorkbook(string filename)
 		    }
 		    else
 		    {
-		        this.mBook.SaveAs(filename, mv, mv, mv, mv, mv, XlSaveAsAccessMode.xlNoChange, mv, mv, mv, mv);
-		    }
+                //this.mBook.Save();
```
Comment (mcavigelli) [inline FileHelpers.ExcelStorage/ExcelStorage.cs]: Please remove this comment, probably remainder of some debugging.

File: FileHelpers.Examples/Examples/13.Excel/10.ExcelCreateAndSave.cs (hunk 2 above)
Comment (markosimulak) [inline FileHelpers.Examples/Examples/13.Excel/10.ExcelCreateAndSave.cs]: Allows to run an example when example form is running (enables button for running following code)

File: FileHelpers.Examples/Examples/ExamplesGenerator.autogen.cs (hunk 6 above)
Comment (markosimulak) [inline FileHelpers.Examples/Examples/ExamplesGenerator.autogen.cs]: Damn. I was looking how to run *.tt but didn't see it.. So I wrote it manually.

**Comment by markosimulak** [comment]
Hi Matthias,

You'r wellcome.

"Unfortunately I was not able to regenerate the examples; could you tell me how to do it?"
I just run <Example project> and on app window select <Excel example>

P.S. I have a problem when need to open excel file (example no. 2). Because application always creating new folder in debug, and ExcelNPOI storage read from that new created folder. So, I need to set a breakpoint in code, go to that new folder and add some excel file.

I fixed it, hope everything ok now.

Thank you for reviewing my code,
Marko

**Comment by mcavigelli** [comment]
Thank you for your two examples.

When running the ExamplesGenerator.tt I still get the following 4 errors
1. `Compiling transformation: 'GeneratedTextTransformation.Write()' must declare a body because it is not marked abstract, extern, or partial`
2. `Compiling transformation: Identifier expected`
3. `Compiling transformation: Invalid token 'this' in class, struct, or interface member declaration`
4. `Compiling transformation: Method must have a return type`

But these errors don't seem to be related to your change.

Matthias

**Comment by markosimulak** [comment]
Again, you're wellcome :)

I'm not sure why you get this errors, because I don't have any error or warning.

Take care,
Marko

**Pull request merged**